planner/core/rule_partition_processor.go

```diff
@@ -328,37 +339,24 @@ func (s *partitionProcessor) pruneRangePartition(ds *DataSource, pi *model.Parti
 	result := fullRange(len(pi.Definitions))
 	// Extract the partition column, if the column is not null, it's possible to prune.
 	if col != nil {
-		// TODO: Store LessThanData in the partitionExpr, avoid allocating here.
-		lessThan, err := makeLessThanData(pi)
+		partExpr, err := ds.table.(partitionTable).PartitionExpr()
 		if err != nil {
 			return nil, err
 		}
-		pruner := rangePruner{lessThan, col, fn}
+		pruner := rangePruner{
+			lessThan: lessThanDataInt{
+				data:     partExpr.ForRangePruning.LessThan,
```

Comment:
Using copy here is better? Copy could avoid race here.

Reply:
It's read-only data... free from data race, so the copy is unnecessary.

The `LessThan` is different from `expression.Expression` or `ast.StmtNode`
The [visit function](https://github.com/pingcap/parser/blob/master/ast/ast.go#L154) modify the node like [here](https://github.com/pingcap/parser/blob/cd90cc2a7d870118b3319bf50564a6d55613c145/ast/expressions.go#L201):

```
node.XX = XX.Accept(Visitor)
``` 

And the expression is also modified during use (ugly code).